server/src/main/java/org/candlepin/dto/DTOFactory.java

```diff
@@ -0,0 +1,259 @@
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.dto;
+
+import org.candlepin.model.ModelEntity;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.HashMap;
+
+
+
+/**
+ * The DTOFactory class provides functionality for building DTOs from other objects. The DTOFactory
+ * works by delegating the translation work to one or more DTOTranslator instances, which are
+ * registered to a given DTOFactory instance.
+ */
+public class DTOFactory {
+    private static Logger log = LoggerFactory.getLogger(DTOFactory.class);
+
+    protected Map<Class, DTOTranslator> translators;
+
+
+    /**
+     * Initializes a new DTOFactory instance.
+     */
+    public DTOFactory() {
+        this.translators = new HashMap<Class, DTOTranslator>();
+    }
+
+    /**
+     * Registers the given translator to the specified model class. If a translator has already
+     * been registered with the given class, the translator will be replaced with the newly
+     * specified translator, and the previous translator will be returned. If the model class has
+     * not yet been registered to a translator, this method returns null.
+     *
+     * @param srcClass
+     *  The model class to register
+     *
+     * @param translator
+     *  The translator to register for the given class
+     *
+     * @throws IllegalArgumentException
+     *  if either srcClass or translator are null
+     *
+     * @return
+     *  The translator previously registered for the given class, or null if the model class had not
+     *  yet been registered.
+     */
+    public <T extends ModelEntity> DTOTranslator registerTranslator(Class<T> srcClass,
+        DTOTranslator<T, ?> translator) {
+
+        if (srcClass == null) {
+            throw new IllegalArgumentException("srcClass is null");
+        }
+
+        if (translator == null) {
+            throw new IllegalArgumentException("translator is null");
+        }
+
+        DTOTranslator existing = this.translators.get(srcClass);
+
+        this.translators.put(srcClass, translator);
+
+        return existing;
+    }
+
+    /**
+     * Unregisters and returns any translator associated with the given class. If the class has not
+     * been associated with a translator, this method returns null.
+     *
+     * @param srcClass
+     *  The model class to unregister
+     *
+     * @throw IllegalArgumentException
+     *  if srcClass is null
+     *
+     * @return
+     *  The previously registered translator for the given class, or null a translator had not been
+     *  registered
+     */
+    public DTOTranslator unregisterTranslator(Class srcClass) {
```

Comment:
On second thought, I don't believe this would work correctly with those parameterized outputs. In the case where we register some interface I for DTOTranslator<I>, then give it a class C that implements I, we would return DTOTranslator<C> instead of DTOTranslator<I>.